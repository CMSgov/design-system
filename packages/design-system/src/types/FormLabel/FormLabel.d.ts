import * as React from 'react';

export type FormLabelComponent = 'label' | 'legend';

export interface FormLabelProps {
  /**
   * Label text or HTML.
   */
  children: React.ReactNode;
  /**
   * Additional classes to be added to the root element.
   */
  className?: string;
  /**
   * The root HTML element used to render the label
   */
  component?: FormLabelComponent;
  /**
   * Enable the error state by providing an error message.
   */
  errorMessage?: React.ReactNode;
  /**
   * Additional classes to be added to the error message
   */
  errorMessageClassName?: string;
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId?: string;
  /**
   * Additional hint text to display
   */
  hint?: React.ReactNode;
  /**
   * A unique `id` for the label element. Useful for referencing the label from
   * other components with `aria-describedby`.
   */
  id?: string;
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed?: boolean;
  /**
   * Text showing the requirement (ie. "Optional", or "Required").
   * In most cases, this should be used to indicate which fields are optional.
   * See the [form guidelines]({{root}}/guidelines/forms/) for more info.
   */
  requirementLabel?: React.ReactNode;
  /**
   * Additional classes to be added to the label text.
   */
  textClassName?: string;
}
export default class FormLabel extends React.Component<
  FormLabelProps & React.ComponentPropsWithRef<'label'> & React.ComponentPropsWithRef<'legend'>,
  any
> {
  render(): JSX.Element;
}
