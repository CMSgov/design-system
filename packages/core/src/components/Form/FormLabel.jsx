import classNames from 'classnames';
import React from 'react';
import PropTypes from 'prop-types';

/**
 * The FormLabel component provides the label/legend for a field, along with any
 * associated hint text and error messaging.
 *
 * TODO(sawyer): Show this in the documentation
 */
export class FormLabel extends React.PureComponent {
  errorMessage() {
    if (this.props.errorMessage) {
      return (
        <span
          className='ds-c-field__hint ds-u-color--error'
          id={`${this.props.fieldId}-message`}
          role='alert'
        >
          {this.props.errorMessage}
        </span>
      );
    }
  }

  hint() {
    if (this.props.hint) {
      const classes = classNames(
        'ds-c-field__hint',
        {'ds-c-field__hint--inverse': this.props.inversed}
      );
      return <span className={classes}>{this.props.hint}</span>;
    }
  }

  render() {
    const ComponentType = this.props.component;
    const labelTextClasses = this.props.errorMessage && 'ds-u-font-weight--bold';

    return (
      <ComponentType className='ds-c-label' htmlFor={this.props.fieldId}>
        <span className={labelTextClasses}>{this.props.children}</span>
        {this.errorMessage()}
        {this.hint()}
      </ComponentType>
    );
  }
}

FormLabel.defaultProps = { component: 'label' };
FormLabel.propTypes = {
  children: PropTypes.node.isRequired,
  /** The root HTML element used to render the label */
  component: PropTypes.oneOf(['label', 'legend']),
  /** Enable the error state by providing an error message. */
  errorMessage: PropTypes.string,
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId: PropTypes.string,
  /**
   * Hint text. Typically this is a string, but you can pass in additional
   * HTML if you need to further format things.
   */
  hint: PropTypes.node,
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed: PropTypes.bool
};

export default FormLabel;
