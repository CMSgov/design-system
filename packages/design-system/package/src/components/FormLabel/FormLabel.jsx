import InlineError from '../InlineError/InlineError';
import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';

export class FormLabel extends React.PureComponent {
  hint() {
    const { hint } = this.props;
    let { requirementLabel } = this.props;
    if (!hint && !requirementLabel) return;

    const classes = classNames('ds-c-field__hint', {
      'ds-c-field__hint--inverse': this.props.inversed,
    });

    let hintPadding = null;

    if (requirementLabel && hint) {
      if (typeof requirementLabel === 'string') {
        // Remove any existing spacing and punctuation
        requirementLabel = requirementLabel.trim().replace(/\.$/, '');
        // Add punctuation after the requirementLabel so it doesn't run into the hint
        requirementLabel = requirementLabel + '.';
      }

      // Add space between hint and preceding requirementLabel
      hintPadding = ' ';
    }

    return (
      <span className={classes}>
        {requirementLabel}
        {hintPadding}
        {hint}
      </span>
    );
  }

  errorMessage() {
    if (this.props.errorMessage) {
      // Include fallback for errorId for usage outside of FormControl
      let errorId = null;
      if (this.props.errorId) {
        errorId = this.props.errorId;
      } else if (this.props.fieldId) {
        errorId = `${this.props.fieldId}-error`;
      }

      return (
        <InlineError
          id={errorId}
          inversed={this.props.inversed}
          className={this.props.errorMessageClassName}
        >
          {this.props.errorMessage}
        </InlineError>
      );
    }
  }

  render() {
    const {
      fieldId,
      id,
      children,
      component,
      hint,
      textClassName,
      className,
      inversed,
      errorMessage,
      errorMessageClassName,
      errorId,
      requirementLabel,
      ...labelProps
    } = this.props;
    const ComponentType = this.props.component;
    const classes = classNames('ds-c-label', className, {
      'ds-c-label--inverse': inversed,
    });

    return (
      <ComponentType className={classes} htmlFor={fieldId} id={id} {...labelProps}>
        <span className={classNames(textClassName)}>{children}</span>
        {this.hint()}
        {this.errorMessage()}
      </ComponentType>
    );
  }
}

FormLabel.defaultProps = { component: 'label' };
FormLabel.propTypes = {
  /**
   * Label text or HTML.
   */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /**
   * Additional classes to be added to the root element.
   */
  className: PropTypes.string,
  /** The root HTML element used to render the label */
  component: PropTypes.oneOf(['label', 'legend']),
  /** Enable the error state by providing an error message. */
  errorMessage: PropTypes.node,
  /**
   * Additional classes to be added to the error message
   */
  errorMessageClassName: PropTypes.string,
  /**
   * The ID of the error message applied to this field.
   */
  errorId: PropTypes.string,
  /**
   * The ID of the field this label is for. This is used for the label's `for`
   * attribute and any related ARIA attributes, such as for the error message.
   */
  fieldId: PropTypes.string,
  /**
   * Additional hint text to display
   */
  hint: PropTypes.node,
  /**
   * A unique `id` for the label element. Useful for referencing the label from
   * other components with `aria-describedby`.
   */
  id: PropTypes.string,
  /**
   * Set to `true` to apply the "inverse" theme
   */
  inversed: PropTypes.bool,
  /**
   * Text showing the requirement (ie. "Optional", or "Required").
   * In most cases, this should be used to indicate which fields are optional.
   * See the [form guidelines]({{root}}/guidelines/forms/) for more info.
   */
  requirementLabel: PropTypes.node,
  /**
   * Additional classes to be added to the label text.
   */
  textClassName: PropTypes.string,
};

export default FormLabel;
