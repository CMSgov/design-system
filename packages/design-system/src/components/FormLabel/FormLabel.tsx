import React from 'react';
import { Label, LabelProps } from '../Label';

type LabelComponentProps = React.ComponentPropsWithRef<'label'> &
  React.ComponentPropsWithRef<'legend'> &
  LabelProps;

export const FormLabel = (props: LabelComponentProps) => {
  if (process.env.NODE_ENV !== 'production') {
    console.warn(
      'Deprecation warning: The `FormLabel` has been renamed to `Label`. The `FormLabel` export will be removed in a future version of the design system.'
    );
  }

  return <Label {...props} />;
};
