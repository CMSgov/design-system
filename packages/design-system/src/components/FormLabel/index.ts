export { default as FormLabel } from './FormLabel';
