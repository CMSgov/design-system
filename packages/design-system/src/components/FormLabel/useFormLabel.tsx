import React from 'react';
import InlineError from '../InlineError/InlineError';
import classNames from 'classnames';
import mergeIds from '../utilities/mergeIds';
import useId from '../utilities/useId';
import { LabelProps } from '../Label';
import { errorPlacementDefault } from '../flags';

// TODO: Reimplement focusTrigger in another place, like another hook

// Some of the FormLabel's props we pass on to the component prop definitions.
// TODO: This could use a better name
type PassedOnFormLabelProps = Omit<
  LabelProps,
  'children' | 'className' | 'component' | 'fieldId' | 'id' | 'errorMessage'
>;

/**
 * This is the set of public-facing props that each component that uses `useFormLabel`
 * can include in its own props definition.
 */
export interface FormFieldProps extends PassedOnFormLabelProps {
  /**
   * Additional classes to be added to the root element.
   */
  className?: string;
  /**
   * The ID of the error message applied to this field.
   */
  errorId?: string;
  /**
   * Location of the error message relative to the field input
   */
  errorPlacement?: 'top' | 'bottom';
  /**
   * Enable the error state by providing an error message.
   */
  errorMessage?: React.ReactNode;
  /**
   * Additional classes to be added to the error message
   */
  errorMessageClassName?: string;
  /**
   * A unique `id` for the field element. Useful for referencing the field from
   * other components with `aria-describedby`.
   */
  id?: string;
  /**
   * Label for the field.
   */
  label: React.ReactNode;
  /**
   * Additional classes to be added to the field label
   */
  labelClassName?: string;
  /**
   * A unique `id` to be used on the field label. If one isn't provided, a unique ID
   * will be generated.
   */
  labelId?: string;
}

/**
 * This is the full list of props accepted by `useFormLabel`. Components that use
 * `useFormLabel` should not include this entire set in their props but instead use
 * `FormFieldProps`, which is the set of public-facing props that a component can
 * include in its own props.
 */
export interface UseFormLabelProps extends FormFieldProps {
  /**
   * The root HTML element used to render the field label
   */
  labelComponent: 'label' | 'legend';
  /**
   *
   */
  wrapperIsFieldset: boolean;
}

/**
 * This hook contains business logic for how an input field relates to its label
 * and wrapping container element. It takes a component's props and generates
 * the specialized set of props for its label, field, wrapper element, and
 * possible detached "bottom error". While the order and structure of those
 * elements are different between components and are specific to their
 * implementation, much of the logic for prop/attribute generation is consistent
 * between components. The hook will return the relevant information as
 * properties of an object:
 *
 *   - labelProps:   Props to be applied to a `FormLabel` (where we get the name)
 *   - fieldProps:   Props to be applied to the field (a.k.a., the input)
 *   - wrapperProps: Props to be applied to the wrapping element
 *   - errorElement: A rendered React element representing a field error
 *   - errorId:      The id (string) of the error-message element, in case we
 *                   need to reference it (currently only necessary to support
 *                   the FormControl component, which this hook aims to replace)
 *
 * Note that while it's helpful now to abstract away this logic behind a shared
 * hook, that may not always be the case. The first step in creating this hook
 * was to allow an alternative to `FormControl` that allowed for more freedom in
 * how the component's individual elements were structured at to get rid of the
 * dichotomy it imposes on components where their implementation must be split
 * between an inner and outer component.
 */
export function useFormLabel<T extends UseFormLabelProps>(props: T) {
  // TODO: Once we're on React 18, we can use the `useId` hook
  const id = useId('field--', props.id);
  const labelId = props.labelId ?? `${id}__label`;
  const errorId = props.errorId ?? `${id}__error`;
  const hintId = props.hintId ?? `${id}__hint`;

  const {
    className,
    label,
    labelClassName,
    labelComponent,
    // Throw away this value and don't pass it to `fieldProps`
    labelId: _labelId,
    errorMessage,
    errorMessageClassName,
    errorPlacement = errorPlacementDefault(),
    hint,
    requirementLabel,
    inversed,
    wrapperIsFieldset,
    ...remainingProps
  } = props;

  const errorElement = errorMessage ? (
    <InlineError
      id={errorId}
      inversed={inversed}
      className={classNames(
        errorMessageClassName,
        errorPlacement === 'bottom' && 'ds-c-inline-error--bottom'
      )}
    >
      {errorMessage}
    </InlineError>
  ) : undefined;
  const topError = errorPlacement === 'top' ? errorElement : undefined;
  const bottomError = errorPlacement === 'bottom' ? errorElement : undefined;
  const ariaInvalid = props['aria-invalid'] ?? !!errorMessage;

  const ariaDescribedBy =
    mergeIds(
      props['aria-describedby'],
      errorElement && errorId,
      (hint || requirementLabel) && hintId
    ) || undefined;

  const labelProps = {
    children: label,
    className: labelClassName,
    component: labelComponent,
    errorMessage: topError,
    errorId,
    // Avoid using `for` attribute for components with multiple inputs
    // i.e. ChoiceList, DateField, and other components that use `fieldset`
    fieldId: wrapperIsFieldset ? undefined : id,
    hint,
    hintId,
    id: labelId,
    requirementLabel,
    inversed,
  };

  const fieldProps = {
    ...remainingProps,
    id,
    errorMessage,
    inversed,
    'aria-describedby': !wrapperIsFieldset ? ariaDescribedBy : undefined,
    'aria-invalid': !wrapperIsFieldset ? ariaInvalid : undefined,
  };

  const wrapperClassNames = classNames({ 'ds-c-fieldset': wrapperIsFieldset }, className);
  const wrapperProps = {
    className: wrapperClassNames,
    'aria-describedby': wrapperIsFieldset ? ariaDescribedBy : undefined,
    'aria-invalid': wrapperIsFieldset ? ariaInvalid : undefined,
  };

  return { labelProps, fieldProps, wrapperProps, bottomError, errorId };
}

export default useFormLabel;
