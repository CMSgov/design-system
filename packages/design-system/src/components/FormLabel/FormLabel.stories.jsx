import React from 'react';

import FormLabel from './FormLabel';

export default {
  title: 'Components/FormLabel',
  component: FormLabel,
  argTypes: {
    errorMessage: {
      control: { type: 'text' },
    },
    hint: {
      control: { type: 'text' },
    },
    requirementLabel: {
      control: { type: 'text' },
    },
  },
  args: {
    children: 'Sample Label',
  },
  parameters: {
    backgrounds: {
      default: 'light',
      values: [
        {
          name: 'light',
          value: '#fff',
        },
        {
          name: 'Hcgov dark',
          value: '#112e51',
        },
        {
          name: 'Mgov dark',
          value: '#146a5d',
        },
      ],
    },
  },
};

const Template = ({ data, ...args }) => <FormLabel {...args} />;

export const FormLabelDefault = Template.bind({});

export const FormLabelWithHint = Template.bind({});
FormLabelWithHint.args = {
  hint: 'Example of a hint for a form label',
};

export const FormLabelWithError = Template.bind({});
FormLabelWithError.args = {
  errorMessage: 'Example of an error for a form label',
};

export const InverseFormLabel = Template.bind({});
InverseFormLabel.args = {
  inversed: true,
};

InverseFormLabel.parameters = {
  backgrounds: { default: process.env.STORYBOOK_DS === 'mgov' ? 'Mgov dark' : 'Hcgov dark' },
};
