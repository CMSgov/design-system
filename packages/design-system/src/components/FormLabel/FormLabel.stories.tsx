import React from 'react';
import FormLabel from './FormLabel';
import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof FormLabel> = {
  title: 'Components/FormLabel',
  component: FormLabel as any,
  args: {
    children: 'Sample Label',
  },
};
export default meta;

type Story = StoryObj<typeof FormLabel>;

export const FormLabelDefault: Story = {};

export const InverseFormLabel: Story = {
  args: {
    inversed: true,
  },
  parameters: {
    // Must supply `layout: 'fullscreen'` when we use `onDark: true`
    onDark: true,
    layout: 'fullscreen',
  },
};
