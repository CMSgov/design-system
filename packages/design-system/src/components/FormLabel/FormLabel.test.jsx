import FormLabel from './FormLabel';
import React from 'react';
import { shallow } from 'enzyme';

describe('FormLabel', () => {
  const labelText = 'Hello world';

  it('renders label text', () => {
    const props = { fieldId: 'name' };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper.text()).toBe(labelText);
    expect(wrapper.is('label')).toBe(true);
    expect(wrapper.prop('htmlFor')).toBe(props.fieldId);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders error state', () => {
    const props = {
      errorMessage: 'Nah, try again.',
      fieldId: 'name'
    };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders hint string', () => {
    const props = { hint: 'President #44' };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders hint node', () => {
    const props = {
      hint: <strong>President #44</strong>
    };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders requirementLabel string', () => {
    const props = {
      requirementLabel: 'Optional'
    };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders requirementLabel node', () => {
    const props = {
      requirementLabel: <em>It is really optional</em>
    };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('adds punctuation to requirementLabel when hint is also present', () => {
    let props = { hint: 'Hint', requirementLabel: 'Optional' };
    let wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();

    props = {
      hint: <span>Hint</span>,
      requirementLabel: <span>Optional</span>
    };
    wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('avoids duplicate punctuation after requirementLabel', () => {
    const props = { hint: 'Hint', requirementLabel: 'Optional.' };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders as a legend element', () => {
    const props = { component: 'legend' };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper.is('legend')).toBe(true);
  });

  it('is inversed', () => {
    const props = {
      hint: 'Foo',
      inversed: true
    };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(
      wrapper
        .find('.ds-c-field__hint')
        .first()
        .hasClass('ds-c-field__hint--inverse')
    ).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('supports additional classNames', () => {
    const props = { className: 'ds-u-foo', labelClassName: 'ds-u-bar' };
    const wrapper = shallow(<FormLabel {...props}>{labelText}</FormLabel>);

    expect(wrapper).toMatchSnapshot();
  });
});
