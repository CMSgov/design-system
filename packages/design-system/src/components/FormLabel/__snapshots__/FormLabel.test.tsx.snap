// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormLabel adds punctuation to requirementLabel when hint is also present 1`] = `
<label
  class="ds-c-label"
>
  <span
    class=""
  >
    Hello world
  </span>
  <span
    class="ds-c-field__hint"
  >
    Optional.
     
    Hint
  </span>
</label>
`;

exports[`FormLabel is inversed 1`] = `
<span
  class="ds-c-field__hint ds-c-field__hint--inverse"
>
  Foo
</span>
`;

exports[`FormLabel renders as a legend element 1`] = `
<legend
  class="ds-c-label"
>
  <span
    class=""
  >
    Hello world
  </span>
</legend>
`;

exports[`FormLabel renders label text 1`] = `
<span
  class=""
>
  Hello world
</span>
`;

exports[`FormLabel renders requirementLabel 1`] = `
<label
  class="ds-c-label"
>
  <span
    class=""
  >
    Hello world
  </span>
  <span
    class="ds-c-field__hint"
  >
    <em>
      It is really optional
    </em>
  </span>
</label>
`;

exports[`FormLabel supports additional attributes 1`] = `
<label
  aria-label="testing aria"
  class="ds-c-label ds-u-foo"
>
  <span
    class=""
  >
    Hello world
  </span>
</label>
`;
